src/org/mirah/typer/typer.mirah

```diff
@@ -574,6 +569,7 @@ class Typer < SimpleNodeVisitor
     end
     enclosing_node = node.findAncestor {|n| n.kind_of?(MethodDefinition) || n.kind_of?(Script)}
     if enclosing_node.kind_of? MethodDefinition
+      return nil if isMethodInBlock(MethodDefinition(enclosing_node)) # return types are not supported currently for methods which act as templates
```

Comment:
> This is because the method will be re-inferred during closure creation right?

No. This is for handling methods which are part of anonymous classes which are currently (mis-)using block syntax, as explained in https://github.com/mirah/mirah/issues/354 .
That's one reason why I think that block syntax and anonymous classes syntax should be separated.

These "special" MethodDefinitions seem to require running Typer#infer() on them (in order to know that they have free variables which should be part of a binding of an enclosing block), but once they have been copied into the AST of the anonymous class, the Typer will run on these MethodDefinitions' clones again, anyway.

If this line was not there, the next line's `infer(enclosing_node)` would result in an endless recursion.